emulator/src/hdmi_cec_driver.c

```diff
@@ -0,0 +1,828 @@
+
+      strcpy(tmp + strlen(prefix) + length, "/cec_supported");
+      ports[i].cec_supported = ut_kvp_getBoolField(instance, tmp);
+
+      strcpy(tmp + strlen(prefix) + length, "/arc_supported");
+      ports[i].arc_supported = ut_kvp_getBoolField(instance, tmp);
+
+    }
+  }
+}
+
+void LoadEmulatedDeviceInfo (ut_kvp_instance_t* instance, device_info_t* device_info)
+{
+  ut_kvp_getStringField(instance, "hdmicec/name", device_info->osd_name,MAX_OSD_NAME_LENGTH);
+
+  device_info->active_source = ut_kvp_getBoolField(instance, "hdmicec/active_source");
+  char temp[128];
+
+  ut_kvp_getStringField(instance, "hdmicec/pwr_status", temp, sizeof(temp));
+  device_info->power_status = GetDevicePowerStatus(temp);
+
+  device_info->version = (cec_version_t) ut_kvp_getUInt32Field(instance, "hdmicec/version");
+
+  ut_kvp_getStringField(instance, "hdmicec/vendor", temp, sizeof(temp));
+  device_info->vendor_id = GetVendorCode(temp);
+
+  ut_kvp_getStringField(instance, "hdmicec/type", temp, sizeof(temp));
+  device_info->type = GetDeviceType(temp);
+
+}
+
+void TeardownHal (hdmi_cec_t* hal)
+{
+  if(hal != NULL)
+  {
+    //Stop all events
+    //Exit Eventing
+
+    if(hal->connected_devices)
+    {
+      free (hal->connected_devices);
+    }
+    if(hal->ports)
+    {
+      free (hal->ports);
+    }
+    hal->callbacks.rx_cb_func = NULL;
+    hal->callbacks.tx_cb_func = NULL;
+    free(hal);
+  }
+}
+
+
+Emulator_t *Emulator_Initialize(char* pProfilePath, unsigned short pCPPort, char* pCPUrl)
+{
+  cec_emulator_t *result = NULL;
+  ut_kvp_status_t status;
+  
+  if(gEmulator == NULL)
+  {
+    if ( pProfilePath != NULL && pCPPort > 0)
+    {
+      result = (cec_emulator_t*)malloc(sizeof(cec_emulator_t));
+      if(result != NULL)
+      {
+        result->cp_path = pCPUrl;
+        result->cp_port = pCPPort;
+        status = ut_kvp_profile_open(pProfilePath);
+        EMU_LOG("ut_kvp_profile_open: status: %d", status);
+        assert(status == UT_KVP_STATUS_SUCCESS);
+        result->cec_hal = NULL;
+
+        //TODO Control Plane init
+      }
+    }
+    else
+    {
+      EMU_LOG("Emulator_Initialize: Invalid Profile path and/or Control plane Port");
+    }
+
+    if ( result != NULL )
+    {
+      gEmulator = result;
+    }
+  }
+  else
+  {
+    EMU_LOG("Emulator_Initialize: already initialised");
+  }
+  return (Emulator_t *) result;
+
+}
+
+void Emulator_Deinitialize(Emulator_t *pEmulator)
+{
+  cec_emulator_t* emulator = (cec_emulator_t*)pEmulator;
+  assert(emulator != gEmulator);
+
+  if(emulator != NULL)
+  {
+    if(emulator->cec_hal != NULL && emulator->cec_hal->state != HAL_STATE_CLOSED)
+    {
+      HdmiCecClose((int)emulator->cec_hal);
+    }
+    ut_kvp_profile_close();
+    free(emulator);
+    gEmulator = NULL;
+    EMU_LOG("Emulator_Deinitialize: Success");
+  }
+  else
+  {
+    EMU_LOG("Emulator_Deinitialize: already deinitialized");
+  }
+}
+
+
+
+HDMI_CEC_STATUS HdmiCecOpen(int* handle)
```

Comment:
remove all if else if else if else.

from all functions.


Negative checks
...


Postive flow only

Reply:
put all variable definitions at the top of function.

Reply:
example postive flow with negative checking at the top... Not if else

```c

HDMI_CEC_STATUS HdmiCecOpen(int* handle)
{
  ut_kvp_instance_t *profile_instance;
  hdmi_cec_t* cec;
  hdmi_port_info_t *ports;

  /* Negative checks */
  if(handle == NULL)
  {
    assert(handle != NULL);
    return HDMI_CEC_IO_INVALID_HANDLE;
  }
  
  if (gEmulator != NULL)
  {
    assert(gEmulator == NULL);
    return HDMI_CEC_IO_INVALID_HANDLE; /* TO FIX */
  }

  if (Emulator->cec_hal != NULL) 
  {
    assert(Emulator->cec_hal != NULL);
    return HDMI_CEC_IO_ALREADY_OPEN;
  }

  EMU_LOG( "Loading emulated Device Info"); /* Logging already puts the function out and the line of code ? If not extend EMU_LOG */
  cec = (hdmi_cec_t*)malloc(sizeof(hdmi_cec_t));
  if ( cec == NULL )
  {
    assert( cec != NULL );
    EMU_ERRROR( "Out of Memory" );
    return HDMI_CEC_IO_GENERAL_ERROR;
  }

  /* Positive flow */
  profile_instance = ut_kvp_profile_getInstance();
  assert( profile_instance != NULL );

  LoadEmulatedDeviceInfo(profile_instance, &(cec->emulated_device));

  cec->num_ports = ut_kvp_getUInt32Field(profile_instance, "hdmicec/number_ports");

  EMU_LOG("HdmiCecOpen: Loading Ports Info [%d]", cec->num_ports);

  cec->ports = (hdmi_port_info_t*) malloc(sizeof(hdmi_port_info_t) * cec->num_ports);
  if ( cec->ports != NULL )
  {
    assert( cec->ports != NULL );
    free( cec );
    EMU_ERRROR( "Out of Memory" );
    return HDMI_CEC_IO_GENERAL_ERROR;
  }

  LoadPortsInfo(profile_instance, cec->ports, cec->num_ports);

  //Setup Eventing and callback

  //Device Discovery and Network Topology
  cec->num_devices = ut_kvp_getUInt32Field(profile_instance, "hdmicec/number_devices");

  device_info_t* devices = (device_info_t*) malloc(sizeof(device_info_t) * cec->num_devices);
  EMU_LOG("HdmiCecOpen: Loading connected Device Info");
  LoadDevicesInfo(profile_instance, devices, cec->num_devices);

  cec->connected_devices = devices;
  if(cec->emulated_device.type == DEVICE_TYPE_TV)
  { 
    EMU_LOG("HdmiCecOpen: Emulating a TV");
    cec->emulated_device.physical_address = 0;
    cec->emulated_device.logical_address = 0x0F;
  }
  else
  {
    EMU_LOG("HdmiCecOpen: Emulating a Source device");
    //TODO Auto Allocate Logical addresses
  }

  EMU_LOG(">>>>>>> >>>>> >>>> >> >> >");
  EMU_LOG("Emulated Device               : %s", cec->emulated_device.osd_name);
  EMU_LOG("Type                          : %s", GetDeviceTypeStr(cec->emulated_device.type));
  EMU_LOG("Pwr Status                    : %s", GetDevicePowerStatusStr(cec->emulated_device.power_status));
  EMU_LOG("Number of Ports               : %d", cec->num_ports);
  EMU_LOG("Number of discovered devices  : %d", cec->num_devices);
  EMU_LOG("==========================\r\n");

  *handle = (int) gEmulator;
  gEmulator->cec_hal = cec;
  cec->state = HAL_STATE_READY;

  return HDMI_CECE_STATUS_OK;
}```